src/main/cpp/profiler/ProfilerJni.cpp

```diff
@@ -0,0 +1,530 @@
+  std::queue<std::unique_ptr<profile_buffer>> buffers_;
+  bool shutdown_ = false;
+};
+
+// Stack of profile buffers that are ready to be filled with profile data.
+struct free_buffer_tracker {
+  explicit free_buffer_tracker(size_t size) : buffer_size_(size) {}
+
+  // Returns the next available profile buffer or creates one if none are available.
+  std::unique_ptr<profile_buffer> get()
+  {
+    {
+      std::lock_guard lock(lock_);
+      if (buffers_.size() > 0) {
+        auto result = std::move(buffers_.top());
+        buffers_.pop();
+        return result;
+      }
+    }
+    return std::make_unique<profile_buffer>(buffer_size_);
+  }
+
+  void put(std::unique_ptr<profile_buffer>&& buffer)
+  {
+    buffer->set_valid_size(0);
+    std::lock_guard lock(lock_);
+    if (buffers_.size() < NUM_CACHED_BUFFERS) {
+      buffers_.push(std::move(buffer));
+    } else {
+      buffer.reset(nullptr);
+    }
+  }
+
+ private:
+  static constexpr size_t NUM_CACHED_BUFFERS = 2;
+  std::mutex lock_;
+  std::stack<std::unique_ptr<profile_buffer>> buffers_;
+  size_t buffer_size_;
+};
+
+void writer_thread_process(JavaVM* vm,
+                           jobject j_writer,
+                           size_t buffer_size,
+                           size_t flush_threshold);
+
+struct subscriber_state {
+  CUpti_SubscriberHandle subscriber_handle;
+  jobject j_writer;
+  std::thread writer_thread;
+  free_buffer_tracker free_buffers;
+  completed_buffer_queue completed_buffers;
+  bool has_cupti_callback_errored;
+  bool is_shutdown;
+
+  subscriber_state(jobject writer, size_t buffer_size)
+    : j_writer(writer),
+      free_buffers(buffer_size),
+      has_cupti_callback_errored(false),
+      is_shutdown(false)
+  {
+  }
+};
+
+// Global variables
+subscriber_state* State = nullptr;
+uint64_t Flush_period_msec;
+uint64_t Last_flush_time_msec;
+
+JavaVM* get_jvm(JNIEnv* env)
+{
+  JavaVM* vm;
+  if (env->GetJavaVM(&vm) != 0) { throw std::runtime_error("Unable to get JavaVM"); }
+  return vm;
+}
+
+JNIEnv* attach_to_jvm(JavaVM* vm)
+{
+  JavaVMAttachArgs args;
+  args.version = JNI_VERSION_1_6;
+  args.name    = const_cast<char*>("profiler writer");
+  args.group   = nullptr;
+  JNIEnv* env;
+  if (vm->AttachCurrentThread(reinterpret_cast<void**>(&env), &args) != JNI_OK) {
+    char const* msg = "PROFILER: unable to attach to JVM";
+    std::cerr << msg << std::endl;
+    throw std::runtime_error(msg);
+  }
+  return env;
+}
+
+char const* get_cupti_error(CUptiResult rc)
+{
+  char const* err;
+  if (cuptiGetResultString(rc, &err) != CUPTI_SUCCESS) { err = "UNKNOWN"; }
+  return err;
+}
+
+void check_cupti(CUptiResult rc, std::string msg)
+{
+  if (rc != CUPTI_SUCCESS) { throw std::runtime_error(msg + ": " + get_cupti_error(rc)); }
+}
+
+uint64_t timestamp_now()
+{
+  timespec info;
+  if (clock_gettime(CLOCK_MONOTONIC_RAW, &info) != 0) {
+    std::cerr << "PROFILER: Unable to determine current time, aborting!" << std::endl;
+    abort();
```

Comment:
nit: I'd rather see an exception if possible, but I don't know if it is possible.

Reply:
The problem with throwing an exception here is that it could be theoretically in any thread, not necessarily a Java-registered thread, and thus the exception will likely end up killing the process.  Given this is just used for periodic flush, I could make this fail in a safer but degraded way, where it simply returns the last flush timestamp and thus effectively disables flushing.